### ISSUE TITLE
FileUpload: Multiple with auto="true" only works first time

### ISSUE THREAD

**Issue by karlkras** [primefaces/primefaces#7364]
**Describe the defect**
Getting irregular and unreliable results when performing an auto, advanced, multiple file upload.  
On startup you simply get a file upload instance to select csv files from. I'm assuming you have, or can get a set of csv files (~4) to be able to use for this. On first selection it appears to work correctly, i.e., select 4 files and upon submit, the table showing files selected will list all four.  
The `Clear List` button does just that... it removes the file selected list and refreshes the section and should see no files selected.  
Now do this again... select the same four files. I'm seeing that subsequent file selections are being missed... in fact, most of the time.

**Reproducer**
I have a fork of the primefaces-test repo [here.](https://github.com/karlkras/primefaces-test)

**Environment:**
 - PF Version: _10.0_
 - JSF + version: _e.g. Mojarra 2.2.20 or MyFaces 2.3.9_
 - Affected browsers: _e.g. Chrome, iOS, ALL_

**To Reproduce**
1. Have 4 csv files available in a test directory.
2. Click on the `Choose` button (FileUpload control)
3. Navigate to directory with test csv files, select all.
4. When updated, you will see list of files selected in a table below the control.
5. (First time in, it appear to show all files correctly.)
6. Click on `Clear List`
7. File resources will be deleted from backing bean, form updated, and fileupload will be available again.
8. Repeat #s 2-4 
9. Subsequent selections will show irregular results in the files that were selected. Mostly incorrectly.

**Expected behavior**
I would expect to see the resulting file list represent exactly what I had selected.

```

**Comment by melloware** [comment]
I have confirmed the problem is `auto="true"`. If you turn `auto="false"` it works every time.

The problem is if Auto its calling upload after each file is added and then submit is immediately called while its still adding files to the collection.

When I step through the debugger I can see files=1 then submitted then files 1 and 2 and submitted etc.  Not sure how to fix it.
```javascript
        for(var i = 0; i < this.files.length; i++) {
            this.files[i].ajaxRequest = this.files[i].row.data('filedata');
            this.files[i].ajaxRequest.submit();
        }
```

**Comment by karlkras** [comment]
If this is the problem, why do I get more than 1 result on some of the attempts?
As mentioned, it seemed to work every time on the very first use.

**Comment by melloware** [comment]
Since its JavaScript its a timing issue so what happens is as each file is uploaded if auto="true it calls `upload` then the 2nd file is added and `upload` is called again.  and the `this.files` collection is also being manipulated by the AJAX request etc.    If you use a debugger which slows everything down you can see the problem every time.   At least from my debugging that seems to be the problem.  Not quite sure why it works the first time it should be sporadic every time.

**Comment by karlkras** [comment]
Also, I am already using the "workaround" but the preference is to not have the need for the extra click.

**Comment by melloware** [comment]
Yep its definitely a bug. Just not sure the correct way to fix it.   Basically `upload` should only be called once after all the files have been processed. But since this same method is used for single file uploads its tricky.

**Comment by karlkras** [comment]
Wouldn't suppose some sort of a timer could be injected to allow all file upload events to complete before firing the update? Sounds sorta hacky, but might be an approach?

**Comment by melloware** [comment]
Yep but typically using Timers is usually a last resort.  The other reason its definitely a timing issue because with auto="false" where all the files are added first then you press Upload it works every time and basically its the same code executing but in a forced order.

**Comment by karlkras** [comment]
And that works because the files are already in memory and the amount of time to perform the actual "upload" is comparatively faster?

**Comment by melloware** [comment]
Its because Upload is only being called once with all 4 files properly in the list.  The other way upload is being called 4 times one for every file and looping over the files list and thus the sporadic results.

**Comment by karlkras** [comment]
I see.

**Comment by melloware** [comment]
I think there something bigger going on here.  even adding a delay waiting 500 milliseconds its still sporadic on the 2nd upload.

**Comment by karlkras** [comment]
@melloware , actually, I'm having the same problem with the fileUpload advanced mode even in auto="false" state.
Basically it looks like the advanced state is technically broken for multi file use.

**Comment by karlkras** [comment]
so this brings me back to my original implementation of having to use standard. Bummer.

**Comment by melloware** [comment]
Can you try this reproducer?
[pf-7364.zip](https://github.com/primefaces/primefaces/files/6555250/pf-7364.zip)

It works for me every time?

**Comment by melloware** [comment]
Oh actually you are right.  Something is broken.

**Comment by karlkras** [comment]
It's interesting that this hasn't been noticed before. I would have thought it would be a rather popular component/state.
We've been using standard mode since moving from richfaces to primefaces 8.0 in our application rewrite early last year when we saw this behavior then. I figured it was a lack of knowledge from the dev doing the work on how the component worked and I'd take another look at it as we were moving to 10.0.0
Looks like it's been in the code like this for some time.

**Comment by melloware** [comment]
I guess people have not been using multiple or that it works the first time and nobody reported it not working.

**Comment by jepsar** [comment]
I did some tests and so far found that:

1. The upload is not guaranteed to work the first time.
2. The issue is not related to `auto="true"`.
3. The issue is not related to `multiple="true"`.
4. After upgrading to jQuery File Upload Plugin 10.31.0 the issue is still present.
5. `sequential="true"` will only upload the first file.
6. `onAdd="console.log(this)"` will block uploading.
7. `fail: function(e, data) {` is not triggered when a file is not sent.
8. Also happens without `update="container of upload"`.
9. Happens on Mojarra and MyFaces 2.2 and 2.3.
10. Happens with other scopes than `ViewScoped`.

**Comment by karlkras** [comment]
`The issue is not related to multiple="true".`
So you're saying this happens even for single file uploads?

**Comment by jepsar** [comment]
Yes, with single you can select 4 files (one at a time, auto false), and the same thing might happen.

**Comment by karlkras** [comment]
Well, at least all of advanced mode is uniformily hosed.
That could be a good thing.
Please tell me standard mode is still reliable.

**Comment by ndrikdarto** [comment]
After upgraded to PF10.0.0, I can not upload file. It is just stoped after drag and drop a file. I don't get this situation on PF 8.0.

I get following message on Chrome:
`jQuery.Deferred exception: Cannot read property 'name' of undefined TypeError: Cannot read property 'name' of undefined
    at HTMLDocument.<anonymous> (http://localhost:8080/lunaha/javax.faces.resource/components.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:4:734)
    at HTMLDocument.dispatch (http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:43090)
    at HTMLDocument.v.handle (http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:41074)
    at Object.trigger (http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:71513)
    at HTMLDocument.<anonymous> (http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:72108)
    at Function.each (http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:2976)
    at S.fn.init.each (http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:1454)
    at S.fn.init.trigger (http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:72084)
    at Object.beforeSend (http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:3:1889)
    at Function.ajax (http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:77957) undefined`

On FireFox:
`bindTriggers/<@http://localhost:8080/lunaha/javax.faces.resource/components.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:4:690
dispatch@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:43090
add/v.handle@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:41074
trigger@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:71513
trigger/<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:72108
each@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:2976
each@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:1454
trigger@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:72084
beforeSend@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:3:1889
ajax@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:77957
f@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:11759
_onSend@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:12631
K.widget/</ai[aq]</<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:1111
_addConvenienceMethods/f.submit@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:8271
upload@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:3:10017
addFileToRow@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:3:6403
add@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:3:2593
_trigger@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:8448
_onAdd/<@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:13518
each@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:2976
_onAdd@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:13356
K.widget/</ai[aq]</<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:1111
_onDrop/<@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:16603
c@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:28294
fireWith@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:29039
e@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:30277
l/</t<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:30307
setTimeout handler*l/<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:30516
c@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:28294
add@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:28572
then/<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:30612
Deferred@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:31025
then@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:30541
_handleFileTreeEntries@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:14618
K.widget/</ai[aq]</<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:1111
_getDroppedFiles@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:14808
K.widget/</ai[aq]</<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:1111
_onDrop@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:16484
K.widget/</ai[aq]</<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:1111
an@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:7023
dispatch@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:43090
add/v.handle@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:41074
EventListener.handleEvent*add@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:41541
ke/<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:40138
each@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:2976
each@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:1454
ke@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:40114
on@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:46741
_on/<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:7224
each@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:3026
_on@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:6862
_initEventHandlers@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:16773
K.widget/</ai[aq]</<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:1111
_create@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:18533
K.widget/</ai[aq]</<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:1111
_createWidget@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:3530
K.widget/K[al][ag]@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:639
K.widget.bridge/K.fn[ah]/<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:2726
each@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:2976
each@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:1454
K.widget.bridge/K.fn[ah]@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery-plugins.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:14:2617
init@http://localhost:8080/lunaha/javax.faces.resource/fileupload/fileupload.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:3:4735
Class.extend/e[d]</<@http://localhost:8080/lunaha/javax.faces.resource/core.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:24:335
c@http://localhost:8080/lunaha/javax.faces.resource/core.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:24:437
createWidget@http://localhost:8080/lunaha/javax.faces.resource/core.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:18:6402
cw@http://localhost:8080/lunaha/javax.faces.resource/core.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:18:6146
@http://localhost:8080/lunaha/app/sales/salesmanagement/saletransactionupload/saletransactionupload.xhtml:4:25
e@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:30005
l/</t<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:30307
setTimeout handler*l/<@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:30516
c@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:28294
fireWith@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:29039
fire@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:29075
c@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:28294
fireWith@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:29039
ready@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:32012
B@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:31791
EventListener.handleEvent*@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:32160
@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:220
@http://localhost:8080/lunaha/javax.faces.resource/jquery/jquery.js.xhtml?ln=primefaces&v=10.0.0&e=10.0.0:2:225`

is it possible that this is the same bug?

**Comment by melloware** [comment]
@ndrikdarto I don't think this is the same issue yours might be this; https://github.com/primefaces/primefaces/issues/7151

**Comment by karlkras** [comment]
So what happens to this? It just gets logged and no further work on it as no one can figure out how to fix it?
This would seem to be a pretty serious issue wouldn't it?

**Comment by tandraschko** [comment]
You can either pay for support, wait until a contributor like i have time or do it by yourself and become a contributor.

**Comment by wal-jan** [comment]
I did some analysis on this issue. I think there are (at least 2) race conditions. First on `UIInput#submittedValue` and second in the test class on `FileUploadView#fileList`. The root cause is that the FileUpload widget does send asynchronous HTTP requests in parallel. These requests are processed by the underlying HTTP layer in **parallel threads**.
In each of the threads

- the submitted value will be decoded, which results in a call to **UIInput#setSubmittedValue(Object)**
  (in this case the submitted value is a UploadedFileWrapper)
- the submitted value will be validated, which gets the submitted value by calling **UIInput#getSubmittedValue()**
  - if the submitted value is null, the validation is done, nothing more to do
  - if the submitted value is valid the FileUploadEvent will be queued
  - if the submitted value is valid the value will be transferred into the value attribute calling UIInput#setValue(Object)
    and the submittedValue will be reset to null calling **UIInput#setSubmittedValue(null)**
- the queued event will be passed to the event handler, which puts the uploaded file into a HashMap calling **HashMap.put(Object, Object)**

The race condition on `FileUploadView#fileList` can be avoided by using `ConcurrentHashMap`instead of `HashMap` (but I do not think thats a good solution). I do not have any clue how to mitigate the race condition on `UIInput#submittedValue`.

**Comment by melloware** [comment]
Good analysis @wal-jan I will let @tandraschko chime in on the race condition.

**Comment by tandraschko** [comment]
i dont think that multi-upload with async requests is a good idea
either send all files in one request - or queue requests

**Comment by karlkras** [comment]
> You can either pay for support, wait until a contributor like i have time or do it by yourself and become a contributor.

Hey, I sponsor many contributors and am very appreciative of the support. I'm just very surprised that (IMO) such  important functionality has gone this long (has this always been broken?) without notice.

**Comment by wal-jan** [comment]
> I'm just very surprised that (IMO) such important functionality has gone this long (has this always been broken?) without notice.

@karlkras after some more analysis I do not think any more that the FileUpload functionality is broken. Let me explain.

- In the reproducer the FileUpload component is backed by a `@ViewScoped` bean. In Primefaces showcase the bean is `@RequestScoped`. The FileUpload component in advanced mode sends each file asynchronously in parallel (see my previous analysis above). In the showcase each request works on a fresh FileUpload instance, so no race conditions can occur. In the reproducer all requests work on the same FileUpload instance.
  
  I think for this use case there have to be a `@RequestScoped FileUploadHandler` and a `@ViewScoped FileUploadView` bean. The FileUploadHandler may `@Inject FileUploadView` to call a threadsafe method.
 
- Because file upload requests are standard Faces requests each of them runs through the JSF lifecycle. Each request updates the model and each response contains the components to be updated. This leads to 2 race conditions in the reproducer
  - `update="test-file-selection"`: after each file upload the panel will be updated. In the end the panel is updated once for each file uploaded. (I think this is sowewhat unexpected here, but imo it is not a bug.)
     As a result the DataTable showing the list of uploaded files may be incomplete even if all files were uploaded since the reponses are not guaranteed to come in same order as requests were sent.
  - `disabled="#{fileUploadView.uploadedFiles.size() gt 0}"`: after the first file upload request has reached the end of PROCESS_VALIDATIONS phase the FileUpload component is set to disabled=true because of the given EL expression. All subsequent file upload requests that have not reached APPLY_REQUEST_VALUES phase aren't even decoded because the component is disabled. (I think this is sowewhat unexpected here too, but imo it is not a bug.)

  I think these 2 race conditions can be avoided by using RemoteCommand
  ```
  <p:remoteCommand name="updateFileSelection"
                   actionListener="#{fileUploadView.setUploadComplete(true)}" update="..."/>
  <p:fileUpload ... widgetVar="fileUpload" ~~update="..."~~
                    onComplete="if ( !PF('fileUpload').files.length ) updateFileSelection()"
                    disabled="#{fileUploadView.uploadComplete}" .../>
  ```
  (`onComplete` fires after each file uploaded, so it must be checked that no more files are left for upload.)

I suggest to add some clarifications to the *Advanced File Upload* documentation section on how to use FileUpload in this mode. Maybe FileUpload can be enhanced in one way or another, e.g. by adding an `onCompleteMulti` attribute which fires only after *all* files have been uploaded.

@jepsar I think all of your findings can be explained with the observed race conditions. Even `sequential=true` works with the suggested changes.

[pf-7364-adjusted.zip](https://github.com/primefaces/primefaces/files/6834947/pf-7364-adjusted.zip)

**Comment by melloware** [comment]
so now that integration tests have been provided and a working test case @karlkras @wal-jan can this issue be closed?

**Comment by karlkras** [comment]
> so now that integration tests have been provided and a working test case @karlkras @wal-jan can this issue be closed?

I see a test case for simple mode, does this apply to auto too? That's where it all started.

**Comment by tandraschko** [comment]
And we need at least docs or really change the behavior

**Comment by wal-jan** [comment]
@karlkras test case for mode=advanced auto=false multiple=true is https://github.com/primefaces/primefaces/blob/171c42d7abc62b0af85e8008d7b4e86cb136dcc4/primefaces-integration-tests/src/main/webapp/fileupload/fileUpload006.xhtml#L17-L18
